### ISSUE TITLE
[BUG] Flaky test failure IndexServiceTests.testAsyncTranslogTrimTaskOnClosedIndex

### ISSUE THREAD

**Issue by mch2** [opensearch-project/OpenSearch#6124]
``
./gradlew ':server:test' --tests "org.opensearch.index.IndexServiceTests.testAsyncTranslogTrimTaskOnClosedIndex" -Dtests.seed=A35A0BBD1D4D543A -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=da-DK -Dtests.timezone=Australia/Darwin -Druntime.java=19
``
https://github.com/opensearch-project/OpenSearch/pull/6122

**Comment by peternied** [comment]
`./gradlew ':server:test' --tests "org.opensearch.index.IndexServiceTests.testAsyncTranslogTrimTaskOnClosedIndex" -Dtests.seed=322535FDADE7B0A4 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=ms -Dtests.timezone=SystemV/MST7 -Druntime.java=19`
https://build.ci.opensearch.org/job/gradle-check/14778/testReport/junit/org.opensearch.index/IndexServiceTests/testAsyncTranslogTrimTaskOnClosedIndex/


- Impacted PR https://github.com/opensearch-project/OpenSearch/pull/7374

**Comment by dbwiddis** [comment]
Unable to reproduce locally.

From Jenkins log, the failing assertion is:
https://github.com/opensearch-project/OpenSearch/blob/2416d4574f251180447bd7c9602a3981b329d010/server/src/test/java/org/opensearch/index/IndexServiceTests.java#L449-L454
```
Expected: <55L>
     but: was <110L>
```
This implies the header may be written twice in some cases.

**Comment by ankitkala** [comment]
I was able to reproduce this only once after running around 10K test iterations. 

- So the test creates an index, ingests few docs and then closes the index. After closing the index we assert that the translog is empty by asserting that `translogStats.getTranslogSizeInBytes` equals `DEFAULT_HEADER_SIZE_IN_BYTES`(55 in this case).
- Translog stats for closed index are getting initialized by creating a new `LocalTranslog` and fetching its stats [here](https://github.com/opensearch-project/OpenSearch/blob/aca2e9de7daaadee77c33e21ff5215c3f1e8600f/server/src/main/java/org/opensearch/index/engine/ReadOnlyEngine.java#L273-L284)
- `LocalTranslog`  computes the size by adding the size of all readers [here](https://github.com/opensearch-project/OpenSearch/blob/09bacee5fc85676e97bee6b4ad87dec35c6aa8cc/server/src/main/java/org/opensearch/index/translog/Translog.java#L466-L469). These reader are getting initialized with translogs from disk [here](https://github.com/opensearch-project/OpenSearch/blob/0a2009250cd857d68cdfa87340f8cd1b906d8c6e/server/src/main/java/org/opensearch/index/translog/LocalTranslog.java#L76)
-  As hinted by @dbwiddis, the translog size being 110 instead of 55(default header size) implies there are 2 empty translog files (on the disk).
- During index close we perform a flush where we rollover the translog and trim unreferenced readers([here](https://github.com/opensearch-project/OpenSearch/blob/c0852f84f0647f78a45e71396d6370a39ba4922f/server/src/main/java/org/opensearch/index/engine/InternalEngine.java#L1863-L1890)). 
- Translog rollover should happen only when the current writer has some operations([here](https://github.com/opensearch-project/OpenSearch/blob/09bacee5fc85676e97bee6b4ad87dec35c6aa8cc/server/src/main/java/org/opensearch/index/translog/Translog.java#L1711-L1715)). So Ideally we shouldn't have 2 empty files on the disk.

**Comment by ankitkala** [comment]
```
[2023-11-21T03:55:00,762][INFO ][o.o.i.IndexService       ] [node_s_0] [test] [0] closing... (reason: [removing index (CLOSED)])
[2023-11-21T03:55:00,762][INFO ][o.o.i.IndexService       ] [node_s_0] [test] [0] closed (reason: [removing index (CLOSED)])
[2023-11-21T03:55:00,775][INFO ][o.o.p.PluginsService     ] [node_s_0] PluginService:onIndexModule index:[test/6Z35y_VQTpmWn0x78xUklQ]
[2023-11-21T03:55:00,787][INFO ][o.o.i.t.LocalTranslog    ] [node_s_0] [test][0] ankikala: Found a reader for /private/var/folders/dz/5106kpws767d85yrxw9y1m040000gr/T/org.opensearch.index.IndexServiceTests_2FC8623D3F0B25BE-001/tempDir-002/data/nodes/0/indices/6Z35y_VQTpmWn0x78xUklQ/0/translog/translog-16.tlog
[2023-11-21T03:55:00,802][INFO ][o.o.c.r.a.AllocationService] [node_s_0] Cluster health status changed from [RED] to [GREEN] (reason: [shards started [[test][0]]]).
[2023-11-21T03:55:00,980][INFO ][o.o.i.t.LocalTranslog    ] [node_s_0] [test][0] ankikala: Found a reader for /private/var/folders/dz/5106kpws767d85yrxw9y1m040000gr/T/org.opensearch.index.IndexServiceTests_2FC8623D3F0B25BE-001/tempDir-002/data/nodes/0/indices/6Z35y_VQTpmWn0x78xUklQ/0/translog/translog-17.tlog
[2023-11-21T03:55:00,981][INFO ][o.o.i.t.LocalTranslog    ] [node_s_0] [test][0] ankikala: Found a reader for /private/var/folders/dz/5106kpws767d85yrxw9y1m040000gr/T/org.opensearch.index.IndexServiceTests_2FC8623D3F0B25BE-001/tempDir-002/data/nodes/0/indices/6Z35y_VQTpmWn0x78xUklQ/0/translog/translog-16.tlog
[2023-11-21T03:55:00,987][WARN ][o.o.i.t.LocalTranslog    ] [node_s_0] [test][0] deleted previously created, but not yet committed, next generation [translog-18.tlog]. This can happen due to a tragic exception when creating a new generation
[2023-11-21T03:55:00,990][WARN ][o.o.i.e.Engine           ] [node_s_0] [test][0] failed engine [translog trimming failed]
org.opensearch.index.translog.TranslogException: Unable to get the earliest last modified time for the transaction log
	at org.opensearch.index.translog.Translog.earliestLastModifiedAge(Translog.java:423) ~[classes/:?]
	at org.opensearch.index.translog.LocalTranslog.stats(LocalTranslog.java:138) ~[classes/:?]
	at org.opensearch.index.engine.NoOpEngine$3.trimUnreferencedTranslogFiles(NoOpEngine.java:211) [classes/:?]
	at org.opensearch.index.shard.IndexShard.trimTranslog(IndexShard.java:1484) [classes/:?]
	at org.opensearch.index.shard.IndexShard$4.doRun(IndexShard.java:2848) [classes/:?]
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:911) [classes/:?]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [classes/:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:840) [?:?]
[2023-11-21T03:55:00,998][INFO ][o.o.i.IndexService       ] [node_s_0] [test] [0] closing... (reason: [shard failure, reason [translog trimming failed]])
[2023-11-21T03:55:00,998][INFO ][o.o.i.IndexService       ] [node_s_0] [test] [0] closed (reason: [shard failure, reason [translog trimming failed]])
[2023-11-21T03:55:00,998][WARN ][o.o.i.s.IndexShard       ] [node_s_0] [test][0] failed to turn off translog retention
org.opensearch.index.engine.EngineException: failed to trim translog
	at org.opensearch.index.engine.NoOpEngine$3.trimUnreferencedTranslogFiles(NoOpEngine.java:229) ~[classes/:?]
	at org.opensearch.index.shard.IndexShard.trimTranslog(IndexShard.java:1484) ~[classes/:?]
	at org.opensearch.index.shard.IndexShard$4.doRun(IndexShard.java:2848) [classes/:?]
	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:911) [classes/:?]
	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [classes/:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
	at java.base/java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: org.opensearch.index.translog.TranslogException: Unable to get the earliest last modified time for the transaction log
	at org.opensearch.index.translog.Translog.earliestLastModifiedAge(Translog.java:423) ~[classes/:?]
	at org.opensearch.index.translog.LocalTranslog.stats(LocalTranslog.java:138) ~[classes/:?]
	at org.opensearch.index.engine.NoOpEngine$3.trimUnreferencedTranslogFiles(NoOpEngine.java:211) ~[classes/:?]
	... 7 more
```


Note: `Found a reader for...` is a log that i had added within `this.readers.addAll(recoverFromFiles(checkpoint))` in LocalTranslog [here](https://github.com/opensearch-project/OpenSearch/blob/0a2009250cd857d68cdfa87340f8cd1b906d8c6e/server/src/main/java/org/opensearch/index/translog/LocalTranslog.java#L76)

**Comment by kotwanikunal** [comment]
@ankitkala Can we create an issue for the unexpected behavior and add in the investigation there?
It will be simpler to keep track of it.

And just so that its documented as well - `NoOpEngine` is used only for tests right?

**Comment by ankitkala** [comment]
> @ankitkala Can we create an issue for the unexpected behavior and add in the investigation there? It will be simpler to keep track of it.
> 
> And just so that its documented as well - `NoOpEngine` is used only for tests right?

Sure. 
NoOpEngine is used when we close a shard.

**Comment by shourya035** [comment]
Reopening since this again failed in https://build.ci.opensearch.org/job/gradle-check/39959/

**Comment by r1walz** [comment]
Closing as already tracked here: #11547
